pmda/density.py

```diff
@@ -0,0 +1,222 @@
+        #                    format="Histogramming %(n_atoms)6d atoms in frame "
+        #                    "%(step)5d/%(numsteps)d  [%(percentage)5.1f%%]\r")
+        # start, stop, step = u.trajectory.check_slice_indices(start, stop, step)
+
+    def _single_frame(self, ts, atomgroups):
+        """
+        Performs computation on single trajectory frame.
+
+        Creates a histogram of positions of atoms in an atom group for a single frame.
+
+        Parameters
+        ----------
+        ts : int
+            current time step
+        atomgroups : AtomGroup
+            atom group for current block
+
+        Returns
+        -------
+
+        """
+        coord = self.current_coordinates(atomgroups[0], self._cutoff, self._atomselection, self._soluteselection, self._update_selection)
+        h, edges = np.histogramdd(coord, bins=self._bins, range=self._arange, normed=False)
+        h = np.array(h)
+        return [h, edges]
+
+    def _conclude(self):
+        """
+        """
+        self._edges = self._results[0, 1]
+        self._grid = self._results[:, 0].sum(axis=0)
+        self._grid /= float(self._n_frames)
+        metadata = self._metadata if self._metadata is not None else {}
+        # metadata['psf'] = u.filename
+        metadata['dcd'] = self._trajectory.filename
+        metadata['atomselection'] = self._atomselection
+        metadata['n_frames'] = self._n_frames
+        metadata['totaltime'] = round(self._n_frames * self._trajectory.dt, 3)
+        metadata['dt'] = self._trajectory.dt
+        metadata['time_unit'] = mda.core.flags['time_unit']
+        try:
+            metadata['trajectory_skip'] = self._trajectory.skip_timestep  # frames
+        except AttributeError:
+            metadata['trajectory_skip'] = 1
+        try:
+            metadata['trajectory_delta'] = self._trajectory.delta  # in native units
+        except AttributeError:
+            metadata['trajectory_delta'] = 1
+        if self._cutoff > 0 and self._soluteselection is not None:
+            metadata['soluteselection'] = self._soluteselection
+            metadata['cutoff'] = self._cutoff  # in Angstrom
+
+        parameters = self._parameters if self._parameters is not None else {}
+        parameters['isDensity'] = False  # must override
+        g = Density(grid=self._grid, edges=self._edges, units={'length': mda.core.flags['length_unit']},
+                    parameters=parameters, metadata=metadata)
+        g.make_density()
+        self.g = g
+        # logger.info("Density completed (initial density in Angstrom**-3)")
+
+    @staticmethod
+    def _reduce(res, result_single_frame):
+        """ 'append' action for a time series"""
+        if isinstance(res, list) and len(res) == 0:
+            res = result_single_frame
+        else:
+            res[0] += result_single_frame[0]
+        return res
+    @staticmethod
+    def current_coordinates(atomgroup, cutoff, atomselection, soluteselection, update_selection):
```

Comment:
probably the whole thing can go and be replaced inline with `group = u.select_atoms(atomselection, updating=update_selection)`